### ISSUE TITLE
[Question] Is prefaulting memory needed?

### ISSUE THREAD

**Issue by shuhaowu** [ros2-realtime-demo/pendulum#90]
I'm looking through the code and I see [this block of code](https://github.com/ros2-realtime-demo/pendulum/blob/dff4bf9263952ab26ef073cc1f6faadbd25d469a/pendulum_utils/src/memory_lock.cpp#L46-L92), which appears to be prefaulting the heap by allocating a non-deterministic amount of RAM until there are no more page faults. I'm not sure what the purpose of this is, given that `mlockall(MCL_CURRENT | MCL_FUTURE)` has been called. Is this needed? Or is this done simply to reserve space in the heap as the code called `mallopt(M_TRIM_THRESHOLD, -1)` and thus no free space is given back to the OS so dynamic allocation can be used later? That doesn't seem like a good idea as the size is not explicitly specified at least with that call? 

What am I missing?

There are also a bit of conflicting advice on the internet: looking at [the PREEMPT_RT docs](https://wiki.linuxfoundation.org/realtime/documentation/howto/applications/memory#stack_memory_for_rt_threads), it says:

> The entire stack of every thread inside the application is forced to RAM when mlockall(MCL_CURRENT) is called. Threads created after a call to mlockall(MCL_CURRENT | MCL_FUTURE) will generate page faults immediately (on creation), as the new stack is immediately forced to RAM (due to the MCL_FUTURE flag). So all RT threads need to be created at startup time, before the RT show time. With mlockall(…) no explicit additional prefaulting necessary to avoid pagefaults during first (or subsequent) access. 

However, the [LWN's realtime developer checklist](https://lwn.net/Articles/837019/) seems to suggest the opposite:

> The third and final step [after mlockall(...)] is to perform pre-faulting, which means causing all page faults ahead of time and having all memory "ready to go". Heap pre-faulting is done by allocating all of the memory that might be needed, then writing a non-zero value to each page to make sure that the writes are not optimized away. Stack pre-faulting should be done in a similar way by creating and writing a big stack frame, something that developers have "learned never to do" in general.

**Comment by shuhaowu** [comment]
After reading a bit, I think I misunderstood what mlockall was doing. If I understand correctly, prefaulting both the stack and the heap should be needed. See https://ogness.net/ese2020/ese2020_johnogness_rtchecklist.pdf.

**Comment by carlossvg** [comment]
>After reading a bit, I think I misunderstood what mlockall was doing. If I understand correctly, prefaulting both the stack and the heap should be needed. See https://ogness.net/ese2020/ese2020_johnogness_rtchecklist.pdf.

Exactly. If the stack or heap grows after `mlockall(MCL_CURRENT | MCL_FUTURE)` is called, a page fault could happen. Pre-faulting is a way to enforce locking all the memory you will use during real-time execution. You can find here a more verbose explanation for the slides you posted above https://youtu.be/NrjXEaTSyrw?t=1319

**Comment by shuhaowu** [comment]
Yeah I ended up watching that video. Thanks again and sorry for the false alarm.

**Comment by shuhaowu** [comment]
Actually, circling back to this again, it seems like if you allocate a large chunk of RAM without accessing it, then call mlockall with MCL_CURRENT, it should lock all memory pages into physical RAM? The man page says:
 
> mlockall()  locks  all  pages mapped into the address space of the calling process.  This includes the pages of the code, data and stack segment, as well as shared libraries, user space kernel data, shared memory, and memory-mapped files.  All mapped pages are guaranteed to  be  resident  in  RAM when the call returns successfully; the pages are guaranteed to stay in RAM until later unlocked.  
>
> MCL_CURRENT Lock all pages which are currently mapped into the address space of the process.

I suppose I should go find out via a test program.

**Comment by shuhaowu** [comment]
Just wrote a [test program](https://github.com/shuhaowu/rt-demo/blob/master/simpler-examples/experiments/prefault-test.cc), the result are the following:

```
Page faults before:                   275, 0
Page faults after malloc + set value: 1315, 0 | +1040, +0
Page faults after malloc only:        2338, 0 | +1023, +0
Page faults after set value:          2338, 0 | +0, +0
```

So in this program, I called `mlockall(MCL_CURRENT | MCL_FUTURE)` first. Then I performed a bunch of mallocs:

1. malloc + writing a value to each page gives a 1000 page fault.
2. malloc without writing a value to each page also gives 1000 faults
3. with the buffer malloc'ed from 2, I then write a value to each page. This gave 0 faults.

This makes sense from the man page documentation. Effectively, we still need to "prefault" the stack/heap, but we don't actually need to write to it (i.e. demand paging has effectively been turned off after mlockall). So the slides [I linked to earlier](https://ogness.net/ese2020/ese2020_johnogness_rtchecklist.pdf) is wrong about that.

**Comment by carlossvg** [comment]
@shuhaowu You are right, when  demand paging is turned off you don't have to touch each page.
Actually this is well explained here:
- https://rt.wiki.kernel.org/index.php/Simple_memory_locking_example
- https://rt.wiki.kernel.org/index.php/Dynamic_memory_allocation_example

In the first case, touching the pages is needed to lock it in memory, in the second case is just a way to demonstrate it is not needed.

**Comment by shuhaowu** [comment]
Yes. It's all documented in the PREEMPT_RT wiki. There's also https://rt.wiki.kernel.org/index.php/Verifying_mlockall()_effects_on_stack_memory_proof which shows that prefaulting the stack is also unnecessary with demand paging off.

That said, I have noticed that various online resources appears to be wrong[^1]. It's good that we sorted it here. I'm working on documenting this in my blog post, and perhaps we should leave a note somewhere in the RT docs, to dispell the confusion for the future.

--------

[^1] For example, the presentation I linked originally has the following code:

```
void prefault_stack(void) { 
  unsigned char dummy[MAX_SAFE_STACK];
  int i;
  for (i = 0; i < MAX_SAFE_STACK; i += sysconf(_SC_PAGESIZE))
    dummy[i] = 1;
}
```

With demand paging off before the thread spawns, this isn't needed. Further, if you look at this function, it basically has no side effect and the compiler could in principle optimize it out (I suppose it can't optimize out `sysconf`? Not sure...). I've also saw other articles and papers saying basically the same, which appears wrong...
